File: backend/src/services/SocketService.js
/**
 * SocketService - сервис для управления WebSocket событиями
 * обрабатывает обновления в реальном времени для заявок
 */

class SocketService {
  /**
   * отправка события создания новой заявки (только пользователям которые могут ее видеть)
   * @param {Object} order - данные созданной заявки
   */
  static async emitOrderCreated(order) {
    if (!global.io) {
      console.warn('📡 [Socket] Cannot emit order:created - io not initialized');
      return;
    }

    console.log('📡 [Socket] Emitting order:created', order.id);

    // отправляем SUPERADMIN и MANAGER (они видят все заявки)
    global.io.to('role:SUPERADMIN').emit('order:created', order);
    global.io.to('role:MANAGER').emit('order:created', order);

    // отправляем владельцу бота (EX_ADMIN) если у заявки есть bot_id
    if (order.bot_id) {
      global.io.to(`bot:${order.bot_id}`).emit('order:created', order);
    }

    // отправляем всем операторам (они теперь видят все заявки, депозит проверяется при взятии)
    global.io.to('operators').emit('order:created', order);
  }

  /**
   * отправка события обновления заявки (тем у кого уже есть доступ)
   * @param {Object} order - обновленные данные заявки
   */
  static emitOrderUpdated(order) {
    if (!global.io) {
      console.warn('📡 [Socket] Cannot emit order:updated - io not initialized');
      return;
    }

    console.log('📡 [Socket] Emitting order:updated', order.id);

    // отправляем SUPERADMIN и MANAGER
    global.io.to('role:SUPERADMIN').emit('order:updated', order);
    global.io.to('role:MANAGER').emit('order:updated', order);

    // отправляем владельцу бота если у заявки есть bot_id
    if (order.bot_id) {
      global.io.to(`bot:${order.bot_id}`).emit('order:updated', order);
    }

    // отправляем назначенному оператору
    if (order.support_id) {
      global.io.to(`user:${order.support_id}`).emit('order:updated', order);
    }
  }

  /**
   * отправка события изменения статуса заявки
   * @param {Object} data - содержит orderId, oldStatus, newStatus, order (полный объект заявки)
   */
  static emitOrderStatusChanged(data) {
    if (!global.io) {
      console.warn('📡 [Socket] Cannot emit order:status-changed - io not initialized');
      return;
    }

    console.log('📡 [Socket] Emitting order:status-changed', data.orderId);

    // отправляем SUPERADMIN и MANAGER
    global.io.to('role:SUPERADMIN').emit('order:status-changed', data);
    global.io.to('role:MANAGER').emit('order:status-changed', data);

    // отправляем владельцу бота если у заявки есть bot_id
    if (data.order?.bot_id) {
      global.io.to(`bot:${data.order.bot_id}`).emit('order:status-changed', data);
    }

    // отправляем назначенному оператору
    if (data.order?.support_id) {
      global.io.to(`user:${data.order.support_id}`).emit('order:status-changed', data);
    }
  }

    /**
   * отправка события взятия заявки (всем кто может видеть заявки)
   * @param {Object} order - данные взятой заявки
   */
  static async emitOrderTaken(order) {
    if (!global.io) {
      console.warn('📡 [Socket] Cannot emit order:taken - io not initialized');
      return;
    }

    console.log('📡 [Socket] Emitting order:taken', order.id);

    // отправляем SUPERADMIN и MANAGER
    global.io.to('role:SUPERADMIN').emit('order:taken', order);
    global.io.to('role:MANAGER').emit('order:taken', order);

    // отправляем владельцу бота
    if (order.bot_id) {
      global.io.to(`bot:${order.bot_id}`).emit('order:taken', order);
    }

    // отправляем всем операторам (они видят все заявки)
    global.io.to('operators').emit('order:taken', order);
  }

  /**
   * отправка события удаления заявки
   * @param {number} orderId - ID удаленной заявки
   */
  static emitOrderDeleted(orderId) {
    if (!global.io) {
      console.warn('📡 [Socket] Cannot emit order:deleted - io not initialized');
      return;
    }

    console.log('📡 [Socket] Emitting order:deleted', orderId);

    // отправляем всем авторизованным пользователям (они отфильтруют на фронтенде)
    global.io.to('role:SUPERADMIN').emit('order:deleted', orderId);
    global.io.to('role:MANAGER').emit('order:deleted', orderId);
    global.io.to('role:EX_ADMIN').emit('order:deleted', orderId);
    global.io.to('operators').emit('order:deleted', orderId);
  }

  /**
   * отправка события сообщения по заявке
   * @param {Object} message - данные сообщения
   */
  static emitOrderMessage(message) {
    if (!global.io) {
      console.warn('📡 [Socket] Cannot emit order:message - io not initialized');
      return;
    }

    console.log('📡 [Socket] Emitting order:message for order', message.order_id);

    // отправляем SUPERADMIN и MANAGER
    global.io.to('role:SUPERADMIN').emit('order:message', message);
    global.io.to('role:MANAGER').emit('order:message', message);

    // отправляем владельцу бота если у заявки есть bot_id
    if (message.bot_id) {
      global.io.to(`bot:${message.bot_id}`).emit('order:message', message);
    }

    // отправляем назначенному оператору
    if (message.support_id) {
      global.io.to(`user:${message.support_id}`).emit('order:message', message);
    }
  }

  /**
   * отправка события подтверждения оплаты пользователем
   * @param {Object} data - содержит order (полный объект заявки), telegramUser (данные пользователя)
   */
  static emitUserPaymentConfirmation(data) {
    if (!global.io) {
      console.warn('📡 [Socket] Cannot emit user:payment-confirmation - io not initialized');
      return;
    }

    console.log('📡 [Socket] Emitting user:payment-confirmation for order', data.order.id);

    global.io.to('role:SUPERADMIN').emit('user:payment-confirmation', data);
    global.io.to('role:MANAGER').emit('user:payment-confirmation', data);

    if (data.order.bot_id) {
      global.io.to(`bot:${data.order.bot_id}`).emit('user:payment-confirmation', data);
    }

    if (data.order.support_id) {
      global.io.to(`user:${data.order.support_id}`).emit('user:payment-confirmation', data);
    }

    global.io.to('operators').emit('user:payment-confirmation', data);
  }

  /**
   * Отправка сообщения в чате оператор <-> менеджер
   * @param {{operator_id:number, manager_id:number, message:Object}} payload
   */
  static emitOperatorManagerMessage(payload) {
    if (!global.io) {
      console.warn('[Socket] Cannot emit operator-manager-chat:message - io not initialized');
      return;
    }

    const operatorId = Number(payload?.operator_id || 0);
    const managerId = Number(payload?.manager_id || 0);
    const message = payload?.message || null;

    const eventPayload = {
      operator_id: operatorId,
      manager_id: managerId,
      message
    };

    if (operatorId > 0) {
      global.io.to(`user:${operatorId}`).emit('operator-manager-chat:message', eventPayload);
    }
    if (managerId > 0) {
      global.io.to(`user:${managerId}`).emit('operator-manager-chat:message', eventPayload);
    }
    global.io.to('role:MANAGER').emit('operator-manager-chat:message', eventPayload);
    global.io.to('role:SUPERADMIN').emit('operator-manager-chat:message', eventPayload);
  }

  /**
   * Отметка сообщений прочитанными в чате оператор <-> менеджер
   * @param {{operator_id:number, manager_id:number, reader_role:string, reader_id:number, marked:number}} payload
   */
  static emitOperatorManagerRead(payload) {
    if (!global.io) {
      console.warn('[Socket] Cannot emit operator-manager-chat:read - io not initialized');
      return;
    }

    const operatorId = Number(payload?.operator_id || 0);
    const managerId = Number(payload?.manager_id || 0);
    const eventPayload = {
      operator_id: operatorId,
      manager_id: managerId,
      reader_role: payload?.reader_role || null,
      reader_id: Number(payload?.reader_id || 0),
      marked: Number(payload?.marked || 0)
    };

    if (operatorId > 0) {
      global.io.to(`user:${operatorId}`).emit('operator-manager-chat:read', eventPayload);
    }
    if (managerId > 0) {
      global.io.to(`user:${managerId}`).emit('operator-manager-chat:read', eventPayload);
    }
    global.io.to('role:MANAGER').emit('operator-manager-chat:read', eventPayload);
    global.io.to('role:SUPERADMIN').emit('operator-manager-chat:read', eventPayload);
  }

  /**
   * Обновление назначения менеджера оператору
   * @param {{operator_id:number, manager_id:number|null}} payload
   */
  static emitOperatorManagerAssignmentUpdated(payload) {
    if (!global.io) {
      console.warn('[Socket] Cannot emit operator-manager-chat:assignment-updated - io not initialized');
      return;
    }

    const eventPayload = {
      operator_id: Number(payload?.operator_id || 0),
      manager_id: payload?.manager_id === null || payload?.manager_id === undefined
        ? null
        : Number(payload.manager_id)
    };

    if (eventPayload.operator_id > 0) {
      global.io.to(`user:${eventPayload.operator_id}`).emit('operator-manager-chat:assignment-updated', eventPayload);
    }
    if (eventPayload.manager_id) {
      global.io.to(`user:${eventPayload.manager_id}`).emit('operator-manager-chat:assignment-updated', eventPayload);
    }
    global.io.to('role:SUPERADMIN').emit('operator-manager-chat:assignment-updated', eventPayload);
    global.io.to('role:MANAGER').emit('operator-manager-chat:assignment-updated', eventPayload);
  }

  /**
   * получение количества подключенных клиентов
   * @returns {Number} количество подключенных клиентов
   */
  static getConnectedClientsCount() {
    if (global.io) {
      return global.io.engine.clientsCount || 0;
    }
    return 0;
  }
}

module.exports = SocketService;

File: python_backend/app/socket/socket_service.py
"""
SocketService analog backend/src/services/SocketService.js

all methods static, get sio (socketio.AsyncServer) through init().
similar commands Node.js versions — frontend not changed:
  role:SUPERADMIN, role:MANAGER, role:EX_ADMIN
  operators
  user:{userId}
  bot:{botId}
"""

import datetime
import decimal
import logging
import socketio

logger = logging.getLogger(__name__)

sio: socketio.AsyncServer | None = None


def init(server: socketio.AsyncServer) -> None:
    global sio
    sio = server


def _s(obj):
    """Recursively convert non-JSON-serializable types for socket emission."""
    if isinstance(obj, dict):
        return {k: _s(v) for k, v in obj.items()}
    if isinstance(obj, (list, tuple)):
        return [_s(v) for v in obj]
    if isinstance(obj, (datetime.datetime, datetime.date)):
        return obj.isoformat()
    if isinstance(obj, decimal.Decimal):
        return float(obj)
    if isinstance(obj, bytes):
        return obj.decode("utf-8", errors="replace")
    return obj


async def emit_order_created(order: dict) -> None:
    if not sio:
        return
    d = _s(order)
    await sio.emit("order:created", d, room="role:SUPERADMIN")
    await sio.emit("order:created", d, room="role:MANAGER")
    if order.get("bot_id"):
        await sio.emit("order:created", d, room=f"bot:{order['bot_id']}")
    await sio.emit("order:created", d, room="operators")


async def emit_order_updated(order: dict) -> None:
    if not sio:
        return
    d = _s(order)
    await sio.emit("order:updated", d, room="role:SUPERADMIN")
    await sio.emit("order:updated", d, room="role:MANAGER")
    if order.get("bot_id"):
        await sio.emit("order:updated", d, room=f"bot:{order['bot_id']}")
    if order.get("support_id"):
        await sio.emit("order:updated", d, room=f"user:{order['support_id']}")


async def emit_order_status_changed(data: dict) -> None:
    """data = {orderId, oldStatus, newStatus, order}"""
    if not sio:
        return
    d = _s(data)
    order = data.get("order") or {}
    await sio.emit("order:status-changed", d, room="role:SUPERADMIN")
    await sio.emit("order:status-changed", d, room="role:MANAGER")
    if order.get("bot_id"):
        await sio.emit("order:status-changed", d, room=f"bot:{order['bot_id']}")
    if order.get("support_id"):
        await sio.emit("order:status-changed", d, room=f"user:{order['support_id']}")


async def emit_order_taken(order: dict) -> None:
    if not sio:
        return
    d = _s(order)
    await sio.emit("order:taken", d, room="role:SUPERADMIN")
    await sio.emit("order:taken", d, room="role:MANAGER")
    if order.get("bot_id"):
        await sio.emit("order:taken", d, room=f"bot:{order['bot_id']}")
    await sio.emit("order:taken", d, room="operators")


async def emit_order_deleted(order_id: int) -> None:
    if not sio:
        return
    for room in ("role:SUPERADMIN", "role:MANAGER", "role:EX_ADMIN", "operators"):
        await sio.emit("order:deleted", order_id, room=room)


async def emit_order_message(message: dict) -> None:
    if not sio:
        logger.warning("[SOCKET] emit_order_message called but sio is None")
        return
    d = _s(message)
    order_id = message.get("order_id")
    support_id = message.get("support_id")
    bot_id = message.get("bot_id")
    logger.info(
        f"[SOCKET] emit order:message order_id={order_id} support_id={support_id} bot_id={bot_id}"
    )
    await sio.emit("order:message", d, room="role:SUPERADMIN")
    await sio.emit("order:message", d, room="role:MANAGER")
    if bot_id:
        await sio.emit("order:message", d, room=f"bot:{bot_id}")
    if support_id:
        await sio.emit("order:message", d, room=f"user:{support_id}")
        logger.info(f"[SOCKET] emitted order:message to user:{support_id}")
    else:
        logger.warning(f"[SOCKET] order:message order_id={order_id} — support_id is None, operator won't receive it")


async def emit_user_payment_confirmation(data: dict) -> None:
    """data = {order, telegramUser}"""
    if not sio:
        return
    d = _s(data)
    order = data.get("order") or {}
    await sio.emit("user:payment-confirmation", d, room="role:SUPERADMIN")
    await sio.emit("user:payment-confirmation", d, room="role:MANAGER")
    if order.get("bot_id"):
        await sio.emit("user:payment-confirmation", d, room=f"bot:{order['bot_id']}")
    if order.get("support_id"):
        await sio.emit("user:payment-confirmation", d, room=f"user:{order['support_id']}")
    await sio.emit("user:payment-confirmation", d, room="operators")


# ---------------------------------------------------------------------------
# чат поддержки (support-chat)
# ---------------------------------------------------------------------------

async def emit_support_chat_message(chat_id: int, message: dict) -> None:
    if not sio:
        return
    await sio.emit("support-chat:message", _s({"chatId": chat_id, "message": message}))


async def emit_support_chat_read(chat_id: int) -> None:
    if not sio:
        return
    await sio.emit("support-chat:read", {"chatId": chat_id})


async def emit_support_chat_typing(chat_id: int, operator_id: int, operator_login: str, is_typing: bool) -> None:
    if not sio:
        return
    await sio.emit("support-chat:typing", {
        "chatId": chat_id,
        "operatorId": operator_id,
        "operatorLogin": operator_login,
        "isTyping": is_typing,
    })


async def emit_support_chat_deleted(chat_id: int) -> None:
    if not sio:
        return
    await sio.emit("support-chat:deleted", {"chatId": chat_id})


# ---------------------------------------------------------------------------
# Чат оператор ↔ менеджер
# ---------------------------------------------------------------------------

async def emit_operator_manager_message(operator_id: int, manager_id: int, message: dict) -> None:
    if not sio:
        return
    payload = _s({"operator_id": operator_id, "manager_id": manager_id, "message": message})
    if operator_id:
        await sio.emit("operator-manager-chat:message", payload, room=f"user:{operator_id}")
    if manager_id:
        await sio.emit("operator-manager-chat:message", payload, room=f"user:{manager_id}")
    await sio.emit("operator-manager-chat:message", payload, room="role:MANAGER")
    await sio.emit("operator-manager-chat:message", payload, room="role:SUPERADMIN")


async def emit_operator_manager_read(operator_id: int, manager_id: int, reader_role: str, reader_id: int, marked: int) -> None:
    if not sio:
        return
    payload = {
        "operator_id": operator_id,
        "manager_id": manager_id,
        "reader_role": reader_role,
        "reader_id": reader_id,
        "marked": marked,
    }
    if operator_id:
        await sio.emit("operator-manager-chat:read", payload, room=f"user:{operator_id}")
    if manager_id:
        await sio.emit("operator-manager-chat:read", payload, room=f"user:{manager_id}")
    await sio.emit("operator-manager-chat:read", payload, room="role:MANAGER")
    await sio.emit("operator-manager-chat:read", payload, room="role:SUPERADMIN")


async def emit_operator_manager_assignment_updated(operator_id: int, manager_id: int | None) -> None:
    if not sio:
        return
    payload = {"operator_id": operator_id, "manager_id": manager_id}
    if operator_id:
        await sio.emit("operator-manager-chat:assignment-updated", payload, room=f"user:{operator_id}")
    if manager_id:
        await sio.emit("operator-manager-chat:assignment-updated", payload, room=f"user:{manager_id}")
    await sio.emit("operator-manager-chat:assignment-updated", payload, room="role:SUPERADMIN")
    await sio.emit("operator-manager-chat:assignment-updated", payload, room="role:MANAGER")




